README
NAME
    Map::Tube::Kazan - Interface to the Kazan Metro Map.

SYNOPSIS
     use Map::Tube::Kazan;

     my $obj = Map::Tube::Kazan->new;
     my $routes_ar = $obj->get_all_routes($from, $to);
     my $line = $obj->get_line_by_id($line_id);
     my $line = $obj->get_line_by_name($line_name);
     my $lines_ar = $obj->get_lines;
     my $station = $obj->get_node_by_id($station_id);
     my $station = $obj->get_node_by_name($station_name);
     my $route = $obj->get_shortest_route($from, $to);
     my $stations_ar = $obj->get_stations($line);
     my $metro_name = $obj->name;
     my $xml_file = $obj->xml;

DESCRIPTION
    It currently provides functionality to find the shortest route between
    the two given nodes.

    For more information about Kazan Map, click here
    <https://en.wikipedia.org/wiki/Kazan_Metro>.

METHODS
  "new"
     my $obj = Map::Tube::Kazan->new;

    Constructor.

    Returns instance of object.

  "get_all_routes" [EXPERIMENTAL]
     my $routes_ar = $obj->get_all_routes($from, $to);

    Get all routes from station to station.

    Returns reference to array with Map::Tube::Route objects.

  "get_line_by_id"
     my $line = $obj->get_line_by_id($line_id);

    Get line object defined by id.

    Returns Map::Tube::Line object.

  "get_line_by_name"
     my $line = $obj->get_line_by_name($line_name);

    Get line object defined by name.

    Returns Map::Tube::Line object.

  "get_lines"
     my $lines_ar = $obj->get_lines;

    Get lines in metro map.

    Returns reference to unsorted array with Map::Tube::Line objects.

  "get_node_by_id"
     my $station = $obj->get_node_by_id($station_id);

    Get station node by id.

    Returns Map::Tube::Node object.

  "get_node_by_name"
     my $station = $obj->get_node_by_name($station_name);

    Get station node by name.

    Returns Map::Tube::Node object.

  "get_shortest_route"
     my $route = $obj->get_shortest_route($from, $to);

    Get shortest route between $from and $to node names. Node names in $from
    and $to are case insensitive.

    Returns Map::Tube::Route object.

  "get_stations"
     my $stations_ar = $obj->get_stations($line);

    Get list of stations for concrete metro line.

    Returns reference to array with Map::Tube::Node objects.

  "name"
     my $metro_name = $obj->name;

    Get metro name.

    Returns string with metro name.

  "xml"
     my $xml_file = $obj->xml;

    Get XML specification of Kazan metro.

    Returns string with XML.

EXAMPLE1
     use strict;
     use warnings;

     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Kazan;

     # Object.
     my $obj = Map::Tube::Kazan->new;

     # Get route.
     my $route = $obj->get_shortest_route(decode_utf8('Яшьлек'), decode_utf8('Горки'));

     # Print out type.
     print "Route: ".encode_utf8($route)."\n";

     # Output:
     # Route: Яшьлек (Центральная линия), Козья слобода (Центральная линия), Кремлёвская (Центральная линия), Площадь Габдуллы Тукая (Центральная линия), Суконная слобода (Центральная линия), Аметьево (Центральная линия), Горки (Центральная линия)

EXAMPLE2
     use strict;
     use warnings;

     use Map::Tube::Kazan;

     # Object.
     my $obj = Map::Tube::Kazan->new;

     # Get XML file.
     my $xml_file = $obj->xml;

     # Print out XML file.
     print "XML file: $xml_file\n";

     # Output like:
     # XML file: .*/kazan-map.xml

EXAMPLE3
     use strict;
     use warnings;

     use Map::Tube::GraphViz;
     use Map::Tube::GraphViz::Utils qw(node_color_without_label);
     use Map::Tube::Kazan;

     # Object.
     my $obj = Map::Tube::Kazan->new;

     # GraphViz object.
     my $g = Map::Tube::GraphViz->new(
             'callback_node' => \&node_color_without_label,
             'driver' => 'neato',
             'tube' => $obj,
     );

     # Get graph to file.
     $g->graph('Kazan.png');

     # Print file.
     system "ls -l Kazan.png";

     # Output like:
     # -rw-r--r-- 1 skim skim 27046 Dec 25 18:23 Kazan.png

EXAMPLE4
     use strict;
     use warnings;

     use Encode qw(encode_utf8);
     use Map::Tube::Kazan;

     # Object.
     my $obj = Map::Tube::Kazan->new;

     # Get lines.
     my $lines_ar = $obj->get_lines;

     # Print out.
     map { print encode_utf8($_->name)."\n"; } sort @{$lines_ar};

     # Output:
     # Центральная линия

EXAMPLE5
     use strict;
     use warnings;

     use Encode qw(decode_utf8 encode_utf8);
     use Map::Tube::Kazan;

     # Arguments.
     if (@ARGV < 1) {
             print STDERR "Usage: $0 line\n";
             exit 1;
     }
     my $line = decode_utf8($ARGV[0]);

     # Object.
     my $obj = Map::Tube::Kazan->new;

     # Get stations for line.
     my $stations_ar = $obj->get_stations($line);

     # Print out.
     map { print encode_utf8($_->name)."\n"; } @{$stations_ar};

     # Output:
     # Usage: __PROG__ line

     # Output with 'foo' argument.
     # Map::Tube::get_stations(): ERROR: Invalid Line Name [foo]. (status: 105) file __PROG__ on line __LINE__

     # Output with 'Центральная линия' argument.
     # Авиастроительная
     # Северный вокзал
     # Яшьлек
     # Козья слобода
     # Кремлёвская
     # Площадь Габдуллы Тукая
     # Суконная слобода
     # Аметьево
     # Горки
     # Проспект Победы

DEPENDENCIES
    File::Share, Map::Tube, Moo, namespace::clean.

SEE ALSO
    Map::Tube
        Core library as Role (Moo) to process map data.

    Task::Map::Tube
        Install the Map::Tube modules.

    Task::Map::Tube::Metro
        Install the Map::Tube concrete metro modules.

REPOSITORY
    <https://github.com/michal-josef-spacek/Map-Tube-Kazan>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    Artistic License

    BSD 2-Clause License

VERSION
    0.08



Changes
0.08

0.07 2023-12-04T16:28:19+01:00
 - Update dependency to Map::Tube@3.62.

0.06 2023-12-04T12:54:06+01:00
 - Fix Makefile.PL use lib '.'.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Move bugtracker to github.
 - Rename example files to better names.
 - Rewrite doc to use =head2.
 - Update Module::Install to 1.21 version.
 - Update author github username.
 - Update copyright years.

0.05 2017-11-23T08:42:54+01:00
 - Add EXAMPLE5 section to doc.
 - Add test for ok_map_functions().
 - Fix Makefile.PL for new version of Perl and no default »use lib '.'«.
 - Implement station order.
 - Improve description for get_lines() in doc.
 - Rewrite XML map to new version with line IDs.
 - Update Module::Install to 1.18 version.
 - Update author name.
 - Update copyright years to 2017.
 - Update to Map::Tube 3.41.

0.04 2015-10-15T23:31:12+02:00
 - Add tests for get_node_by_id() and get_node_by_name() methods.
 - Update to Map::Tube 3.06.

0.03 2015-09-27T18:59:14+02:00
 - Add <?xml> header to XML document.
 - Add example for print of metro lines to doc.
 - Add get_line_by_id() and get_line_by_name() methods to doc.
 - Add lines to XML document.
 - Add test of map validity.
 - Fix minimal Perl version.
 - Move author tests to xt/ directory.
 - Rewrite SEE ALSO section to new style.
 - Update copyright years to 2015.
 - Update Module::Install to 1.16 version.

0.02 2014-12-25T18:26:55+01:00
 - Add EXAMPLE3 section to doc.
 - Add Map::Tube::GraphViz, Map::Tube::Text::Table, Map::Tube::Bucharest,
   Map::Tube::Dnipropetrovsk, Map::Tube::Novosibirsk, Map::Tube::Prague,
   Map::Tube::SaintPetersburg, Map::Tube::Tbilisi, Map::Tube::Vienna and
   Map::Tube::Yekaterinburg to SEE ALSO section in doc.
 - Add get_lines() method to doc.
 - Add get_stations() method to doc.
 - Update README.
 - Remove executable bit from kazan-map.xml file.
 - Remove obsolete code from EXAMPLE2 section in doc.

0.01 2014-12-11T07:59:22+01:00
 - First version


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'Interface to the Kazan Metro Map.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.08;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
install_share 'share';
license 'bsd';
license 'artistic_2';
name 'Map-Tube-Kazan';
readme_from 'Kazan.pm';
recursive_author_tests('xt');
requires 'File::Share' => 0;
requires 'Map::Tube' => '3.62';
requires 'Moo' => 0;
requires 'namespace::clean' => 0;
requires 'perl' => '5.6.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Map-Tube-Kazan/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Map-Tube-Kazan';
resources 'repository' => 'git://github.com/michal-josef-spacek/Map-Tube-Kazan';
test_requires 'Encode' => 0;
test_requires 'English' => 0;
test_requires 'Test::Map::Tube' => 0.35;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.08';

# Run.
WriteAll();


Kazan.pm
package Map::Tube::Kazan;

use strict;
use warnings;
use 5.006;

use File::Share ':all';
use Moo;
use namespace::clean;

our $VERSION = 0.08;

# Get XML.
has xml => (
	'is' => 'ro',
	'default' => sub {
		return dist_file('Map-Tube-Kazan', 'kazan-map.xml');
	},
);

with 'Map::Tube';

1;

__END__

=encoding utf8

=head1 NAME

Map::Tube::Kazan - Interface to the Kazan Metro Map.

=head1 SYNOPSIS

 use Map::Tube::Kazan;

 my $obj = Map::Tube::Kazan->new;
 my $routes_ar = $obj->get_all_routes($from, $to);
 my $line = $obj->get_line_by_id($line_id);
 my $line = $obj->get_line_by_name($line_name);
 my $lines_ar = $obj->get_lines;
 my $station = $obj->get_node_by_id($station_id);
 my $station = $obj->get_node_by_name($station_name);
 my $route = $obj->get_shortest_route($from, $to);
 my $stations_ar = $obj->get_stations($line);
 my $metro_name = $obj->name;
 my $xml_file = $obj->xml;

=head1 DESCRIPTION

It currently provides functionality to find the shortest route between the two
given nodes.

For more information about Kazan Map, click L<here|https://en.wikipedia.org/wiki/Kazan_Metro>.

=head1 METHODS

=head2 C<new>

 my $obj = Map::Tube::Kazan->new;

Constructor.

Returns instance of object.

=head2 C<get_all_routes> [EXPERIMENTAL]

 my $routes_ar = $obj->get_all_routes($from, $to);

Get all routes from station to station.

Returns reference to array with Map::Tube::Route objects.

=head2 C<get_line_by_id>

 my $line = $obj->get_line_by_id($line_id);

Get line object defined by id.

Returns Map::Tube::Line object.

=head2 C<get_line_by_name>

 my $line = $obj->get_line_by_name($line_name);

Get line object defined by name.

Returns Map::Tube::Line object.

=head2 C<get_lines>

 my $lines_ar = $obj->get_lines;

Get lines in metro map.

Returns reference to unsorted array with Map::Tube::Line objects.

=head2 C<get_node_by_id>

 my $station = $obj->get_node_by_id($station_id);

Get station node by id.

Returns Map::Tube::Node object.

=head2 C<get_node_by_name>

 my $station = $obj->get_node_by_name($station_name);

Get station node by name.

Returns Map::Tube::Node object.

=head2 C<get_shortest_route>

 my $route = $obj->get_shortest_route($from, $to);

Get shortest route between $from and $to node names. Node names in $from and $to are case insensitive.

Returns Map::Tube::Route object.

=head2 C<get_stations>

 my $stations_ar = $obj->get_stations($line);

Get list of stations for concrete metro line.

Returns reference to array with Map::Tube::Node objects.

=head2 C<name>

 my $metro_name = $obj->name;

Get metro name.

Returns string with metro name.

=head2 C<xml>

 my $xml_file = $obj->xml;

Get XML specification of Kazan metro.

Returns string with XML.

=head1 EXAMPLE1

=for comment filename=print_kazan_route.pl

 use strict;
 use warnings;

 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Kazan;

 # Object.
 my $obj = Map::Tube::Kazan->new;

 # Get route.
 my $route = $obj->get_shortest_route(decode_utf8('Яшьлек'), decode_utf8('Горки'));

 # Print out type.
 print "Route: ".encode_utf8($route)."\n";

 # Output:
 # Route: Яшьлек (Центральная линия), Козья слобода (Центральная линия), Кремлёвская (Центральная линия), Площадь Габдуллы Тукая (Центральная линия), Суконная слобода (Центральная линия), Аметьево (Центральная линия), Горки (Центральная линия)

=head1 EXAMPLE2

=for comment filename=print_kazan_def_xml_file.pl

 use strict;
 use warnings;

 use Map::Tube::Kazan;

 # Object.
 my $obj = Map::Tube::Kazan->new;

 # Get XML file.
 my $xml_file = $obj->xml;

 # Print out XML file.
 print "XML file: $xml_file\n";

 # Output like:
 # XML file: .*/kazan-map.xml

=head1 EXAMPLE3

=for comment filename=print_kazan_image.pl

 use strict;
 use warnings;

 use Map::Tube::GraphViz;
 use Map::Tube::GraphViz::Utils qw(node_color_without_label);
 use Map::Tube::Kazan;

 # Object.
 my $obj = Map::Tube::Kazan->new;

 # GraphViz object.
 my $g = Map::Tube::GraphViz->new(
         'callback_node' => \&node_color_without_label,
         'driver' => 'neato',
         'tube' => $obj,
 );

 # Get graph to file.
 $g->graph('Kazan.png');

 # Print file.
 system "ls -l Kazan.png";

 # Output like:
 # -rw-r--r-- 1 skim skim 27046 Dec 25 18:23 Kazan.png

=begin html

<a href="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-Kazan/master/images/Kazan.png">
  <img src="https://raw.githubusercontent.com/michal-josef-spacek/Map-Tube-Kazan/master/images/Kazan.png" alt="Казанский метрополитен" width="300px" height="300px" />
</a>

=end html

=head1 EXAMPLE4

=for comment filename=print_kazan_lines.pl

 use strict;
 use warnings;

 use Encode qw(encode_utf8);
 use Map::Tube::Kazan;

 # Object.
 my $obj = Map::Tube::Kazan->new;

 # Get lines.
 my $lines_ar = $obj->get_lines;

 # Print out.
 map { print encode_utf8($_->name)."\n"; } sort @{$lines_ar};

 # Output:
 # Центральная линия

=head1 EXAMPLE5

=for comment filename=print_kazan_line_stations.pl

 use strict;
 use warnings;

 use Encode qw(decode_utf8 encode_utf8);
 use Map::Tube::Kazan;

 # Arguments.
 if (@ARGV < 1) {
         print STDERR "Usage: $0 line\n";
         exit 1;
 }
 my $line = decode_utf8($ARGV[0]);

 # Object.
 my $obj = Map::Tube::Kazan->new;

 # Get stations for line.
 my $stations_ar = $obj->get_stations($line);

 # Print out.
 map { print encode_utf8($_->name)."\n"; } @{$stations_ar};

 # Output:
 # Usage: __PROG__ line

 # Output with 'foo' argument.
 # Map::Tube::get_stations(): ERROR: Invalid Line Name [foo]. (status: 105) file __PROG__ on line __LINE__

 # Output with 'Центральная линия' argument.
 # Авиастроительная
 # Северный вокзал
 # Яшьлек
 # Козья слобода
 # Кремлёвская
 # Площадь Габдуллы Тукая
 # Суконная слобода
 # Аметьево
 # Горки
 # Проспект Победы

=head1 DEPENDENCIES

L<File::Share>,
L<Map::Tube>,
L<Moo>,
L<namespace::clean>.

=head1 SEE ALSO

=over

=item L<Map::Tube>

Core library as Role (Moo) to process map data.

=item L<Task::Map::Tube>

Install the Map::Tube modules.

=item L<Task::Map::Tube::Metro>

Install the Map::Tube concrete metro modules.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Map-Tube-Kazan>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

Artistic License

BSD 2-Clause License

=head1 VERSION

0.08

=cut
